pinot-core/src/main/java/com/linkedin/pinot/core/operator/docvalsets/UnSortedSingleValueSet.java

```diff
@@ -44,6 +46,19 @@ public DataType getValueType() {
   }
 
   @Override
+  public void getStringValues(int[] inDocIds, int inStartPos, int inDocIdsSize, String[] outValues, int outStartPos) {
+    ReaderContext context = sVReader.createContext();
+    int inEndPos = inStartPos + inDocIdsSize;
+    try {
+      for (int i = inStartPos; i < inEndPos; i++) {
+        outValues[outStartPos++] = sVReader.getString(inDocIds[i], context);
```

Comment:
what is the thread safety issue here ? Are you suggesting getting a duplicate buffer from reader and then reading single values here

My suggestion is to push the bulk get method down to the reader for performance under different threads. You can also optimize it knowing the chunk boundaries